File: _jaewon/14888.py
# N은 11까지
# 연산자 개수는 10개까지
# 전체 경우의 수를 다 돌아도 시간초과에 걸리지 않음.
import collections
import math

N = int(input())
elements = list(map(int, input().split()))

# +, -, *, //
left_operators = list(map(int, input().split()))
op_map = {0: "+", 1: "-", 2: "*", 3: "//"}

def calculate(operators_array):
    operators = operators_array[:]
    stack = collections.deque(reversed(elements))

    while len(stack) > 1:
        element1 = stack.pop()
        element2 = stack.pop()

        op = operators.pop(0)

        if(op == '+'):
            stack.append(element1 + element2)
        elif(op == '-'):
            stack.append(element1 - element2)
        elif(op == '*'):
            stack.append(element1 * element2)
        elif(op == '//'):
            if (element1 < 0 and element2 < 0):
                stack.append(math.fabs(element1) // math.fabs(element2))
            elif (element1 < 0):
                stack.append((math.fabs(element1) // element2) * -1)
            elif (element2 < 0):
                stack.append((element1 // math.fabs(element2)) * -1)
            else:
                stack.append(element1 // element2)

    result = stack.pop()
    return result

maximum = -10**9
minimum = 10**9

def backtrack(depth, operators):
    global maximum
    global minimum
    if(depth == N-1):
        result = calculate(operators)
        maximum = max(maximum, result)
        minimum = min(minimum, result)
        return

    tmp = operators[:]
    # +, -, *, //
    for index, count in enumerate(left_operators):
        if (count > 0):
            left_operators[index] -= 1
            operator = op_map[index]
            tmp.append(operator)

            backtrack(depth + 1, tmp)

            left_operators[index] += 1
            tmp.pop(-1)

backtrack(0, [])
print(int(maximum))
print(int(minimum))

File: _chongmyoung/14888.java
import java.util.*;

public class Main {
    static int N;                 // 숫자의 개수
    static int[] numbers;         // 숫자 배열
    static int[] operators = new int[4]; // +, -, *, / 연산자의 개수
    static int max = Integer.MIN_VALUE; // 최댓값 저장
    static int min = Integer.MAX_VALUE; // 최솟값 저장

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        N = sc.nextInt();
        numbers = new int[N];
        for (int i = 0; i < N; i++) {
            numbers[i] = sc.nextInt();
        }
        for (int i = 0; i < 4; i++) {
            operators[i] = sc.nextInt(); // + - * / 개수 입력
        }

        dfs(1, numbers[0]); // 첫 번째 숫자부터 시작
        System.out.println(max);
        System.out.println(min);
    }

    // DFS로 연산자 배치 탐색
    static void dfs(int depth, int result) {
        // 모든 숫자를 다 사용한 경우
        if (depth == N) {
            max = Math.max(max, result);
            min = Math.min(min, result);
            return;
        }

        // 각 연산자 종류별로 탐색
        for (int i = 0; i < 4; i++) {
            if (operators[i] > 0) { // 남은 연산자가 있을 경우
                operators[i]--;     // 해당 연산자 사용

                switch (i) {
                    case 0: // +
                        dfs(depth + 1, result + numbers[depth]);
                        break;
                    case 1: // -
                        dfs(depth + 1, result - numbers[depth]);
                        break;
                    case 2: // *
                        dfs(depth + 1, result * numbers[depth]);
                        break;
                    case 3: // /
                        dfs(depth + 1, result / numbers[depth]);
                        break;
                }

                operators[i]++; // 백트래킹
            }
        }
    }
}
